### DISCUSSION TITLE
Morphology - fine-grained features

### DISCUSSION THREAD

**Discussion by k-sap** [explosion/spaCy#9648] [Help: Other Questions] [answered]
I'd like to ask about fine-grained morphological features.

For part of speech, as described in https://spacy.io/api/token, there is `token.pos_` and `token.tag_`. Respectively: coarse-grained (UD-based universal tagset) and fine-grained (language-dependent part of speech tagset).

However, description of `token.morph` (https://spacy.io/api/morphology#morphanalysis) does not provide such distinction (fine-grained/coarse-grained) for morphological features.

Is there any way of these language-specific morphological features from spacy output which I've missed? If there is no official way, maybe it's still possible to get these tags as some kind of by-product.

For example in Polish:
"xpos": "praet:sg:m1:imperf" - `praet` as fine-grained POS and additional morphological features - sg:m1:imperf Id'like to get.
"feats": "Animacy=Hum|Aspect=Imp|Gender=Masc|Mood=Ind|Number=Sing|Tense=Past|VerbForm=Fin|Voice=Act"  - standard morphological features

**Comment by adrianeboyd** [comment, marked as answer]
If you have fine-grained tags saved as `token.tag` that could potentially map to other features like `token.morph`, you can add rules that do this mapping in an `attribute_ruler` component. If you want to see an example, we do this in several trained pipelines like `en_core_web_sm`. We don't have this stored anywhere in a better format, but you can check out `nlp.get_pipe("attribute_ruler").patterns` to see the raw patterns. An example of one pattern is:

```none
  {
    "patterns":[
      [
        {
          "TAG":"-LRB-"
        }
      ]
    ],
    "attrs":{
      "POS":"PUNCT",
      "MORPH":"PunctSide=Ini|PunctType=Brck"
    },
    "index":0
  }
```

If you have the mapping in the spacy v2 tag map or morph rules format, you can import those with the attribute ruler methods, too: https://spacy.io/api/attributeruler#load_from_tag_map

But the mapping will come from an external source outside spacy itself, either from some source like the UD docs or written by hand.

**Reply by k-sap** [reply]
I have [a second similar question](https://github.com/explosion/spaCy/discussions/11354).

For a given language, I can get a set of possible tags by:
```
nlp = spacy.load('en_core_web_lg')
components = {name: component for name, component in nlp.components}
morphological_features = components['morphologizer'].labels
```

However, for some languages, e.g. for the English, there is no `morphologizer` components in the pipeline. Is there a way to get it for English?

**Reply by polm** [reply]
Please don't ask the same question in multiple places; I have followed up in the new thread.
